data/src/clone.rs

```diff
@@ -0,0 +1,70 @@
+// SPDX-FileCopyrightText: 2025 TriliTech <contact@trili.tech>
+//
+// SPDX-License-Identifier: MIT
+
+//! Cloning functionality
+
+/// Persistent cloneable state
+///
+/// This trait is used to clone the persistent state of a value.
+///
+/// Our PVM data types sometimes contain ephemeral state that helps with running things faster.
+/// However, a regular clone of such a type would also clone the ephemeral state, which is not
+/// always desirable. This trait allows to clone only the persistent state, thereby discarding any
+/// ephemeral state.
+pub trait CloneState {
+    /// Clone only the persistent state contained in this value.
+    fn clone_state(&self) -> Self;
+}
+
+impl<T: CloneState, const N: usize> CloneState for [T; N] {
+    fn clone_state(&self) -> Self {
+        self.each_ref().map(CloneState::clone_state)
+    }
+}
+
+// We don't implement `Box<T>` directly, to avoid bringing potentially large `T`s onto the stack.
```

Comment:
In implementing it for `Box<T>`, why would something like `CloneState::clone_state(self.as_ref())` bring the `T` into the stack?